cpp/src/parquet/arrow/path_internal_test.cc

```diff
@@ -0,0 +1,426 @@
+    EXPECT_TRUE(null_rep_levels_);
+    ASSERT_FALSE(null_def_levels_);
+    EXPECT_THAT(def_levels_, ElementsAreArray(expected_def));
+  }
+
+  void CheckLevels(const std::vector<int16_t>& expected_def,
+                   const std::vector<int16_t>& expected_rep) const {
+    ASSERT_FALSE(null_def_levels_);
+    ASSERT_FALSE(null_rep_levels_);
+    EXPECT_THAT(def_levels_, ElementsAreArray(expected_def));
+    EXPECT_THAT(rep_levels_, ElementsAreArray(expected_rep));
+  }
+};
+
+struct Callback {
+  Status operator()(const MultipathLevelBuilderResult& result) {
+    results->emplace_back(result);
+    return Status::OK();
+  }
+  std::vector<CapturedResult>* results;
+};
+
+class MultipathLevelBuilderTest : public testing::Test {
+ protected:
+  std::vector<CapturedResult> results_;
+  Callback callback_{&results_};
+  std::shared_ptr<ArrowWriterProperties> arrow_properties_ =
+      default_arrow_writer_properties();
+  ArrowWriteContext context_ =
+      ArrowWriteContext(default_memory_pool(), arrow_properties_.get());
+};
+
+TEST_F(MultipathLevelBuilderTest, NonNullableSingleListNonNullableEntries) {
+  auto entries = field("Entries", ::arrow::int64(), /*nullable=*/false);
+  auto list_type = large_list(entries);
+  auto array = ::arrow::ArrayFromJSON(list_type, R"([[1], [2, 3], [4, 5, 6]])");
+
+  ASSERT_OK(
+      MultipathLevelBuilder::Write(*array, /*nullable=*/false, &context_, callback_));
+
+  ASSERT_THAT(results_, SizeIs(1));
+  const CapturedResult& result = results_[0];
+
+  result.CheckLevels(/*def_levels=*/std::vector<int16_t>(/*count=*/6, 1),
+                     /*rep_levels=*/{0, 0, 1, 0, 1, 1});
+
+  ASSERT_THAT(result.post_list_elements, SizeIs(1));
+  EXPECT_THAT(result.post_list_elements[0].start, Eq(0));
+  EXPECT_THAT(result.post_list_elements[0].end, Eq(6));
+}
+
+TEST_F(MultipathLevelBuilderTest, NullableSingleListWithAllNullsLists) {
+  auto entries = field("Entries", ::arrow::int64(), /*nullable=*/false);
+  auto list_type = list(entries);
+  auto array = ::arrow::ArrayFromJSON(list_type, R"([null, null, null, null])");
+
+  ASSERT_OK(
+      MultipathLevelBuilder::Write(*array, /*nullable=*/true, &context_, callback_));
+
+  ASSERT_THAT(results_, SizeIs(1));
+  const CapturedResult& result = results_[0];
+  result.CheckLevels(/*def_levels=*/std::vector<int16_t>(/*count=*/4, 0),
+                     /*rep_levels=*/std::vector<int16_t>(4, 0));
+}
+
+TEST_F(MultipathLevelBuilderTest, NullableSingleListWithAllNullEntries) {
+  auto entries = field("Entries", ::arrow::int64(), /*nullable=*/true);
+  auto list_type = list(entries);
+  auto array = ::arrow::ArrayFromJSON(list_type, R"([[null], [null], [null], [null]])");
+
+  ASSERT_OK(
+      MultipathLevelBuilder::Write(*array, /*nullable=*/true, &context_, callback_));
+
+  ASSERT_THAT(results_, SizeIs(1));
+  const CapturedResult& result = results_[0];
+  result.CheckLevels(/*def_levels=*/std::vector<int16_t>(/*count=*/4, 2),
```

Comment:
It might be worth commenting the 3-level encoding for the benefit of a less well-versed reader

```
optional group bag {
  repeated group [unseen] (List) {
    optional int64 Entries;
  }
}
```

We have then

* def level 0: a null list
* def level 1: an empty list
* def level 2: a null entry
* def level 3: a non-null entry

Since this schema is used in multiple tests it may help others understand the expected results